### DISCUSSION TITLE
How to pass "args" to Minimizer.brute?

### DISCUSSION THREAD

**Discussion by aeiou6630** [lmfit/lmfit-py#870] [Q&A] [answered]
Just a quick question. While `scipy.optimize.brute` can take additional arguments via `args = ...`  , I found `Minimizer.brute` doesn't accept any args. Is there a way to add it? Thanks!

**Comment by newville** [comment]
@aeiou6630 looks like we don't support that.

**Reply by aeiou6630** [reply]
Got it. I tried to modify the code in minimizer, but didn't succeed, yet afterwards I was able to use global variables to bypass this issue. Thanks for your reply.

**Comment by newville** [comment]
@aeiou6630 well, we might want to fix this ;)

**Reply by reneeotten** [reply]
please give an example so that it is clear what you mean. The brute function accepts all arguments that the version is scipy does, at least as far as I can tell. If you mean that you want to give a dictionary of them and pass them as **args ; indeed that doesn't work and is by design you have to pass the arguments specifically.

**Reply by newville** [reply]
@reneeotten I think the question is how to pass some object through lmfit into the `args` argument of `scipy.optimize.brute()`.  It is the sort of thing that `fit_kws` should probably handle but I think we do not do this for our wrapping of `brute`.  

Trying to have a coherent mapping of the `scipy.optimize` routines is sort of a continual struggle!

**Reply by reneeotten** [reply]
I am not behind a computer right now so cannot check easily. But from memory: all keywords that scipy.brute accept are part of the functional call in minimizer.brute and can be supplied in the usual way. Again, it's possible that I don't understand the concern here correctly but I'd recommend to look at the examples on how to pass arguments.

**Reply by aeiou6630** [reply]
Let me just copy the source code

(1) `lmfit.minimizer.brute`:
Input: `def brute(self, params=None, Ns=20, keep=50, workers=1, max_nfev=None):`
The kws sent to scipy_brute: `brute_kws = dict(full_output=1, finish=None, disp=False, Ns=Ns, workers=workers)`
Calling scipy_brute: `ret = scipy_brute(self.penalty, tuple(ranges), **brute_kws)`

(2) `scipy.optimize.brute`:
`scipy.optimize.brute(func, ranges, args=(), Ns=20, full_output=0, finish=<function fmin>, disp=False, workers=1)`

I thought there's not an option to pass the `args` variable from `lmfit.minimizer.brute` to `scipy.optimize.brute`, so eventually I used global variables to bypass this issue.

**Reply by reneeotten** [reply]
@aeiou6630 this doesn't show what you are trying to do and state that doesn't work. As I said before, post a minimal example that shows your problem. But again, I don't think there is a problem based on what you copied here; just provide the named arguments to the brute method and it should just work.

**Reply by newville** [reply]
@reneeotten I do agree with @aeiou6630 that there is a problem.  We call `scipy.optimize.brute` with

```
ret = scipy_brute(self.penalty, tuple(ranges), **brute_kws)
```

with `brute_kws` set as 

```
brute_kws = dict(full_output=1, finish=None, disp=False, Ns=Ns, workers=workers)
```

and that is not updated from any attributes of `Minimizer`.   So, there is currently no mechanism for the user to pass in values to the `args` argument of `scipy.optimize.brute`.     We could (perhaps "should") add `args` if the user has passed that into `fit_kws`. 

Sort of a separate topic but there is also no way to override `finish` -- we just don't allow a polishing step.    I think that is fine (we don't want `brute` calling that as it would not be a "Parameter-enabled residual"), but we may want to add code to automatically call a polishing step on N (say 10?) best solutions found by `brute`.

**Reply by reneeotten** [reply]
okay, let me take a look then once I am behind a computer.

Regarding the "polish" step, again that's by design. The optimal solutions found by the "brute" method are saved in an attribute and these can be used as input for a "final" refinement in the usual way (as documented and shown in the example). I don't think that should change as in the current approach all optimizers we support in lmfit can be used.

**Reply by aeiou6630** [reply]
Thanks guys. Though I guess you probably have well understood it, I'll provide a minimal example:

`from scipy.optimize import brute`

`def func(x, x0):`
`    return (x-x0)**2`

`x0 = 5`
`ranges = [slice(0,10,1)]`
`args = tuple([x0])`
`result = brute(func, ranges, args = args)`

, where `x0` is the thing I think `Minimizer.brute` cannot handle at the moment.

**Reply by reneeotten** [reply]
Please look at the [example provided in the gallery](https://lmfit.github.io/lmfit-py/examples/example_brute.html#sphx-glr-examples-example-brute-py). 

For example, there we created a `Minimizer` and provide parameters:
```
fitter = Minimizer(fcn2min, params, fcn_args=(x, data))
result_brute = fitter.minimize(method='brute', Ns=25, keep=25)
```

and checking the `call_kws` that are given in the end to `scipy.optimize.brute` we see: 
```
{'full_output': 1, 'finish': None, 'disp': False, 'Ns': 25, 'workers': 1}
```

exactly as it should be.

So again, I am sorry if I am not understanding it correctly... and as I asked earlier provide an example that *uses* `lmfit` to show what you want to do and what doesn't work.

**Reply by newville** [reply]
@reneeotten Yes, that is the `call_kws`, but this does not include an argument called `args` that `scipy.optimize.brute` supports.  And there is no way to pass that to `lmfit.Minimizer.brute` or into `lmfit.Minimizer` and then have it sent to `scipy.optimize.brute`.

**Reply by reneeotten** [reply, marked as answer]
@newville I don't think it has to include an "args" argument as that is supposed to provide things/parameters to calculate the residual function. That is what "fnc_args" is intended for. The parameter "Ns" in my example was changed from its default and does get passed correctly "scipy.optimize.brute". 

A few other things that are listed in "brute_kws" (i.e., 'full_output', 'finish', 'disp') can indeed not be changed as is documented in the docstring, but that's what I decided on when adding this method. There should be no compelling reason for the end-user to change those.

**Reply by newville** [reply]
@reneeotten Yes, I agree that one could pass `fcn_args` to `lmfit.minimize`.   

It is not always clear how or when to decide which arguments of the `scipy.optimize` routines we support.    
If someone is familiar with  `scipy.optimize.brute` and wants to use the `args` argument, should we allow that?  
Perhaps that should set `self.userargs`,  or maybe it is better to raise a Warning pointing to `fcn_args` to better ensure uniformity in our API (uniformity in `scipy.optimize` appears hopeless to me).   It seems like we currently ignore the argument, which is probably not ideal.

**Reply by aeiou6630** [reply]
Oh, I didn't notice the `fcn_args` option when creating a `Minimizer` object. That indeed solves it. Thanks!

**Comment by reneeotten** [comment]
@newville I see your point and realize now that the `brute` method is the only one that doesn't have a `**kws`` kitchen-sink option for the user to supply whatever they want... I still think that would be okay here as it is documented which options cannot be changed (and I don't see a reason for someone to do that). In most examples (including the ones for `brute`) we use the `fcn_args` argument to the `Minimizer` class to supply the required variable for the residual function instead of using the `args` option that pretty much all `scipy.optimize` solvers have. 

To me that would be a better way of doing it and I'd be in favor of removing the `**kws` option from all solvers we have in `lmfit` and only have the named arguments to the function specifically listed in the function call. However, I realize that this will not fly well as it breaks backwards-compatibility for a not-good-enough reason ;) So if you really prefer I suppose we could add to the `brute` method as well; the only issue is then that we would probably need to remove the "named" arguments as otherwise there is a possibility to provide the same argument twice. Doing so now would of course break backwards compatibility for the `brute` method; anyway, I don't really care - so if you have a strong opinion let me know...

**Comment by newville** [comment]
@reneeotten perhaps we should give a warning if `args` is used with `brute`, suggesting `fcn_args` as the alternative?

We should not feel obligated to follow the `scipy.optimize` API, or really even use their code.  Like, it would be perfectly reasonable to just use our own implementation of `brute` (we've done others in the past).  I'm not advocating for that: we can use their implementations but I think it is OK to say we use them as they suit our needs.  We don't need the `args` argument of `brute`. But maybe we could warn when the user passes in keywords that we know we are going to ignore?
